### PR TITLE
Add `cluster instances migrate replace-voyager` command

### PR DESCRIPTION
**Pull request by matt-mazzucato**
Provide a way for migrating an AstarteVoyagerIngress to an equivalent AstarteDefaultIngress.
The command involves some interactions with a user to gather the information needed to perform some preliminary checks and the actual migration.

This PR falls in the context of https://github.com/astarte-platform/astarte-kubernetes-operator/issues/263 and https://github.com/astarte-platform/astarte/issues/613

### CODE REVIEW

**Review by Annopaolo — CHANGES_REQUESTED** [review]

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+	if err != nil {
+		return err
+	}
+	operatorNamespace, err := command.Flags().GetString("operator-namespace")
+	if err != nil {
+		return err
+	}
+
+	// is the migration allowed?
+	allowed, err := isMigrationAllowed(operatorName, operatorNamespace)
```
[the remark this answers is not in the document]
Comment (matt-mazzucato) [reply cmd/cluster/instance_migrate_to_default_ingress.go]: Changed to `ensureOperatorMinimumVersionRequirement`. How does it sound?

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+	Short:   "Migrate the AstarteVoyagerIngress resource to an equivalent AstarteDefaultIngress",
+	Long:    "Migrate the AstarteVoyagerIngress resource to an equivalent AstarteDefaultIngress",
+	Example: `  astartectl cluster migrate-ingress --namespace astarte --operator-name astarte-operator-controller-manager --operator-namespace kube-system -o avi.bkp`,
+	RunE:    migrateIngressF,
+}
+
+func init() {
+	migrateIngressCmd.PersistentFlags().String("operator-name", "astarte-operator-controller-manager", "The name of the Astarte Operator instance.")
+	migrateIngressCmd.PersistentFlags().String("operator-namespace", "kube-system", "The namespace in which the Astarte Operator resides.")
+	migrateIngressCmd.PersistentFlags().StringP("out", "o", "", "The name of the file in which the AVI custom resource will be saved.")
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate_to_default_ingress.go:46]: AVI --> AstarteVoyagerIngress

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+		return err
+	}
+	if len(aviList.Items) == 0 {
+		return errors.New("No AstarteVoyagerIngress resources are present in the cluster. Nothing to do here.")
+	}
+
+	// assume there is only one avi in the cluster
+	aviObject := &aviList.Items[0]
+
+	fmt.Printf("You are about to migrate the Astarte Voyager Ingress named: %s. ", aviObject.GetName())
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate_to_default_ingress.go:88]: Astarte Voyager Ingress --> AstarteVoyagerIngress

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+	serviceType, found, err := unstructured.NestedString(aviObj.Object, "spec", "api", "type")
+	if err != nil {
+		return ret, err
+	}
+	if !found {
+		// not found, it means the default value (LoadBalancer) is used
+		serviceType = "LoadBalancer"
+	}
+	if serviceType != "LoadBalancer" {
+		fmt.Println(`WARNING: your AstarteVouyagerIngress API configuration falls outside of the supported configurations.
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate_to_default_ingress.go:359]: AstarteVouyagerIngress --> AstarteVoyagerIngress

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+		}
+	}
+
+	// host
+	host, found, err := unstructured.NestedString(aviObj.Object, "spec", "dashboard", "host")
+	if err != nil {
+		return ret, err
+	}
+	if !found {
+		return ret, errors.New("Dashboard host is empty. Something is wrong. Ensure to review ypur Astarte Voyager Ingress configuration.")
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate_to_default_ingress.go:476]: ypur Astarte Voyager Ingress --> your AstarteVoyagerIngress

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+
+	return ret, nil
+}
+
+func maybeNotifyForUnmanagedAnnotations(aviObj *unstructured.Unstructured) {
+	apiAs, apiFound, _ := unstructured.NestedMap(aviObj.Object, "spec", "api", "annotationsService")
+	brokerAs, brokerFound, _ := unstructured.NestedMap(aviObj.Object, "spec", "broker", "annotationsService")
+
+	if apiFound || brokerFound {
+		fmt.Printf("\n-- ATTENTION REQUIRED --\nThe following annotationsService for AstarteVoyagerIngress are found:\n")
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate_to_default_ingress.go:543]: What about `The following AstarteVoyagerIngress annotations have been found`?

**Review by Annopaolo — CHANGES_REQUESTED** [review]
LGTM, just some typos

File: cmd/cluster/instance_migrate_to_default_ingress.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Annopaolo) [reply cmd/cluster/instance_migrate_to_default_ingress.go]: Good!

**Review by rbino — CHANGES_REQUESTED** [review]

File: cmd/cluster/instance_migrate_to_default_ingress.go [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+			fmt.Printf("%s resource successfully written to %s\n", aviObject.GetName(), fn)
+		}
+	}
+
+	adiName, err := utils.PromptChoice("Choose the new AstarteDefaultIngress name:", "adi", false, false)
+	if err != nil {
+		return err
+	}
+
+	// We are not checking if the secret is present in the cluster: if the secret is not present, the validation webhook will return an error
```
Comment (rbino) [inline cmd/cluster/instance_migrate_to_default_ingress.go:112]: Is this valid for both following secrets? If so, use `secrets` instead of `the secret`

**Review by rbino — CHANGES_REQUESTED** [review]

File: cmd/cluster/instance_migrate.go [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+
+var replaceVoyagerCmd = &cobra.Command{
+	Use:   "replace-voyager",
+	Short: "Migrate the AstarteVoyagerIngress resource to an equivalent AstarteDefaultIngress",
+	Long: `Migrate the AstarteVoyagerIngress resource to an equivalent AstarteDefaultIngress.
+
+The user is required to interactively prompt information such as the TLS certificate names and the AstarteDefaultIngress resource name.
+Additionally, for backup purposes, the --out flag allows to dump the AstarteVoyagerIngress resource before starting the migration procedure.
+Before the actual migration starts, the user is required to review the to-be-installed AstarteDefaultIngress resource. The actual migration is performed only upon confirmation.`,
+	Example: `  astartectl cluster instances migrate-ingress --namespace astarte --operator-name astarte-operator-controller-manager --operator-namespace kube-system --out avi-bkp.yaml`,
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate.go:48]: `migrate-ingress` --> `migrate replace-voyager`

**Review by Annopaolo — CHANGES_REQUESTED** [review]
Just one minor leftover

File: cmd/cluster/instance_migrate.go (hunk 8 above)
Comment (matt-mazzucato) [reply cmd/cluster/instance_migrate.go]: Thanks!

File: cmd/cluster/instance_migrate.go [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+		}
+	}
+
+	adiName, err := utils.PromptChoice("Choose the new AstarteDefaultIngress name:", "adi", false, false)
+	if err != nil {
+		return err
+	}
+
+	// We are not checking if the secrets are present in the cluster: if they are not, the validation webhook will return an error
+	apiSecretName, err := utils.PromptChoice("Insert the name of the secret containing the TLS certificates and keys to connect to the Astarte API and Dashboard:", "", false, false)
```
Comment (Annopaolo) [inline cmd/cluster/instance_migrate.go:142]: A user may arrive here after having just created the `astarte-tls-secret` using the [Handling Astarte certificates](https://docs.astarte-platform.org/1.0/050-handling_certificates.html) guide. We should display the `apiSecretName` as default value, as we do in the next step for the broker

**Review by Annopaolo — CHANGES_REQUESTED** [review]

File: cmd/cluster/instance_migrate.go (hunk 9 above)
Comment (matt-mazzucato) [reply cmd/cluster/instance_migrate.go:142]: We have no way of knowing in advance which is the secret to be used and therefore we can not suggest a reasonable default.

**Review by bettio — APPROVED** [review]

**Pull request merged**